Mollic

brownag opened the pull request:
**New functions and documentation pertaining to identification of diagnostic features and properties in SoilProfileCollections:**
- hasDarkColors, mollic.thickness.requirement
- depthOf. maxDepthOf, minDepthOf
- getCambicBounds
- guessHzAttrName, guessHzDesgnName, guessHzTexClName

**Experimental function for modifying horizon boundaries based on a "template" pedon and a vector of boundary distinctness standard deviations:**
- permute_profile

ALSO -- resolves long standing bug #89 and provides several enhancements to i and j indexing of SoilProfileCollections #135

Passing tests on R 3.6.3 (Windows + soilDB w/ local NASIS data )and R 4.0.0 (Ubuntu)

dylanbeaudette commented:
All tests pass on MacOS with R 3.6.1. 

Any reason for replacing `slot(x, 'site')` with `@` method? 

I'd like to look a little closer tonight, otherwise there is a huge pile of functionality in there, wow!

brownag commented:
RE: slot

"In the case of the slot function, name can be any expression that evaluates to a valid slot in the class definition. Generally, the only reason to use the functional form rather than the simpler operator is because the slot name has to be computed."

It is comparable in speed -- I will revert to using slot if you want ... I personally would prefer to use `@` internally wherever possible.

> In `estimateSoilDepth.R` I see that the error for bad horizon names has been removed and replaced with `guessHzDesgnName()`. What happens when `guessHzDesgnName()`fails to find a reasonable name and results in `NA` passed along? I like the approach, but there is no fail-safe now. I'd suggest a test for this condition and further checking / error in `estimateSoilDepth()`. Happy to help test this out.

Mmm that is indeed an oversight. Not sure why I wanted to allow it return `max(NULL)`. I fixed the guess functions so their output is a little more sensible in the pathological unguessable cases... but  less output is generated overall. There were a couple oversights in the flow of that function and `guessHzAttrName`.

> I like all of the new functions (e.g. `mollic.thickness.requirement`), but I do think that we should consider some naming conventions as we move forward. I know there is a mixture of dots and camelCase. I tend to prefer camelCase but don't really care. Lets discuss briefly.

I have generally been following the convention that my highest level functions use camel case, whereas lower level ones use dots. I.e. `getArgillicBounds` `crit.clay.argillic` `argillic.clay.increase.depth` `get.increase.matrix`. I suppose I just made that standard up from observing noise in the patterns of existing function naming... ha

> Looking over the test for `estimateSoilDepth()`, I don't recall that it returns `-Inf`. Am I reading this test correctly? I do think that there should be some kind of warning or error if the user specifies a bogus name to `estimateSoilDepth`. We should not automaticagically fix all input.

Test has been updated to reflect the error -- similar to that used to be generated when `name` was missing. 

```
library(aqp)
x <- data.frame(id=1, top=0:9, bottom=1:10, none=rep(NA, 10), one=rep(1, 10))
depths(x) <- id ~ top + bottom
plot(x)
estimateSoilDepth(x[1,])
```

```
> x <- data.frame(id=1, top=0:9, bottom=1:10, none=rep(NA, 10), one=rep(1, 10))
> depths(x) <- id ~ top + bottom
> plot(x)
unable to guess column containing horizon designations
> estimateSoilDepth(x[1,])
unable to guess column containing horizon designations
Error: soil depth estimation relies on a column containing horizon designations
```

ALSO, I would like to point out that I fixed S4 method dispatch for all SoilProfileCollection methods -- especially those that overlap with `dplyr` -- `aqp` should play nicer now if one is using both at same time -- regardless of which you load first.

dylanbeaudette commented:
Thanks for addressing those. I'm fine with using `@` for slot access, less typing. Function naming is fine with me although I've got a couple of functions that I'd like to rename eventually. Thanks for fixing S4 dispatch. I'll test all of this on my mac with the latest R today.
